src/PureScript/Backend/Optimizer/Tracer/Printer.purs

```diff
@@ -0,0 +1,516 @@
+
+printIndentedParens' :: Doc Void -> Doc Void -> Doc Void
+printIndentedParens' header body =
+  D.lines
+    [ D.text "(" <> header
+    , D.indent body
+    , D.text ")"
+    ]
+
+wrapInParens :: Doc Void -> Doc Void
+wrapInParens = append (D.text "(") <<< flip append (D.text ")")
+
+printLet :: String -> Doc Void -> Doc Void -> Doc Void
+printLet letKeyword identifier binding =
+  D.lines
+    [ D.words
+        [ D.text letKeyword
+        , identifier
+        , D.text "="
+        ]
+    , D.indent $ binding <> D.text ";"
+    ]
+
+printArray :: Array BackendExpr -> Doc Void
+printArray = finish <<< start
+  where
+  finish topPart = D.lines [ topPart, D.text "]" ]
+  start = flip foldlWithIndex mempty \idx acc expr ->
+    if idx == 0 then
+      D.lines
+        [ D.text "["
+        , D.indent $ printBackendExpr expr <> D.text ","
+        ]
+    else
+      D.lines
+        [ acc
+        , D.indent $ printBackendExpr expr <> D.text ","
+        ]
+
+printRecord :: Array (Prop BackendExpr) -> Doc Void
+printRecord = finish <<< start
+  where
+  finish topPart = D.lines [ topPart, D.text "}" ]
+  start = flip foldlWithIndex mempty \idx acc (Prop label expr) ->
+    if idx == 0 then
+      D.lines
+        [ D.words [ D.text "{", D.text label, D.text "=" ]
+        , D.indent $ D.indent $ printBackendExpr expr
+        ]
+    else
+      D.lines
+        [ acc
+        , D.words [ D.text ",", D.text label, D.text "=" ]
+        , D.indent $ D.indent $ printBackendExpr expr
+        ]
+
+printBranch :: NonEmptyArray (Pair BackendExpr) -> BackendExpr -> Doc Void
+printBranch conds fallback = do
+  let
+    printCond idx (Pair cond expr) =
+      D.lines
+        [ D.words
+            [ D.text $ if idx == 0 then "if" else "else if"
+            , D.indent $ printBackendExpr cond
+            , D.text "then"
+            ]
+        , D.indent $ printBackendExpr expr
+        ]
+  D.lines
+    [ D.lines $ Array.mapWithIndex printCond (NonEmptyArray.toArray conds)
+    , D.text "else"
+    , D.indent $ printBackendExpr fallback
+    ]
+
+printBackendSyntax :: BackendSyntax BackendExpr -> Doc Void
+printBackendSyntax = case _ of
+  Var qi ->
+    printQualifiedIdent qi
+  Local mbIdent lvl ->
+    printLocal mbIdent lvl
+  Lit lit ->
+    printLiteral lit
+  App fn args ->
+    (NonEmptyArray.toArray args) # flip Array.foldl (printBackendExpr fn) \acc next ->
+      printCurriedApp acc $ printBackendExpr next
+  Abs args body -> do
+    let
+      printArg arg =
+        D.words
+          [ D.text "\\" <> uncurry printLocal arg
+          , D.text "->" <> D.space
+          ]
+    printIndentedParens' (foldMap1 printArg args) $ printBackendExpr body
+  UncurriedApp fn args ->
+    wrapInParens $ D.words $ map printBackendExpr $ Array.cons fn args
+  UncurriedAbs args body ->
+    printIndentedParens'
+      (D.text "\\" <> (D.words $ map (uncurry printLocal) args) <> D.space <> D.text "->")
+      $ printBackendExpr body
```

Comment:
Just noting here that `Fn0` outputs
```
\ ->
  body
```
I'm not sure whether this is desirable or not.